File: Solutions/80-Remove Duplicates from Sorted Array II/80-Remove Duplicates from Sorted Array II.cs
//Fourth fastest
public class Solution {
    public int RemoveDuplicates(int[] nums) {
        var i=0;
        
        for(var j=0; j<nums.Length; j++)
        {
            if(j<2 || nums[j]>nums[i-2])
            {
                nums[i++] = nums[j];     
            }
        }
        
        return i;
    }
}

//Third fastest
public class Solution {
    public int RemoveDuplicates(int[] nums) {
            if (nums.Length < 3)
            {
                return nums.Length;
            }

            int n = 0;
            int j = 0;
            var tmp = new int[nums.Length];
            for (int i = 0; i < nums.Length; i++)
            {
                if (i == 0)
                {
                    tmp[j] = nums[i];
                    j++;
                }

                if (((i + 1) < nums.Length) && (nums[i] == nums[i + 1]))
                {
                    if (i == 0)
                    {
                        tmp[1] = nums[1];
                        j++;
                        n = 2;
                    }
                    else if (n < 2)
                    {
                        tmp[j] = nums[i];
                        j++;
                        n++;
                    }
                }
                else
                {
                    n = 1;
                   if ((i + 1) < nums.Length)
                    {
                        tmp[j] = nums[i + 1];
                        j++;
                    }
                }
            }
            for (int k = 0; k < j; k++)
            {
                nums[k] = tmp[k];
            }

            return j;
    }
}

//Second fastest
public class Solution {
    public int RemoveDuplicates(int[] nums) {
        
        int skippedItems = 0;
        int i = 2;
        
        while (i < nums.Length - skippedItems)
        {         
            if (skippedItems>0)
                swap(nums, i, i + skippedItems);
            
            if (nums[i] == nums[i-2])
               skippedItems++;
            else
               i++;
            
            
                
        }
        
        return nums.Length-skippedItems;
    }
    
    private void swap(int[] nums, int index, int newIndex)
    {
        var aux = nums[newIndex];
        nums[newIndex] = nums[index];
        nums[index] = aux;
    }
}

//Fastest
public class Solution {
    public int RemoveDuplicates(int[] nums) 
    {
       if(nums.Length == 0) return 0;

            int i = 0;

            int currCount = 0;

            for(int j = 1; j < nums.Length; j++)
            {
                if(nums[i] != nums[j])
                {
                    currCount = 0;
                }
                else
                {
                    currCount++;
                }

                if(currCount < 2)
                {
                    nums[++i] = nums[j];
                }
            }

            return i + 1;
    }
}

File: Solutions/80-Remove Duplicates from Sorted Array II/80-Remove Duplicates from Sorted Array II.py
#Fourth fastest
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
      count, i = 1, 1
      while i < len(nums):
        count = count + 1 if nums[i] == nums[i-1] else 1
        if count > 2: nums.remove(nums[i])
        else: i += 1
      # print(nums)
      return i

#Third fastest
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        check_index = 0
        i = 1
        count = 0
        
        while i < len(nums):
            if nums[check_index] == nums[i]:
                count += 1
                
                if count > 1:
                    nums.remove(nums[i])
                    count -= 1
                    i -= 1
                i += 1
            else:
                check_index = i
                i = check_index + 1
                count = 0            
          
        return len(nums)

#Second fastest
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        j = 0
        count = 0
        for i in range(1, len(nums)):
            if nums[i] == nums[j]:
                count += 1
                if count < 2:
                    j += 1
                    nums[j], nums[i] = nums[i], nums[j]
            else:
                j += 1
                nums[j], nums[i] = nums[i], nums[j]
                count = 0
        return j+1

#Fastest
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        j, count = 1,1
        
        for i in range(1, len(nums)):
            if nums[i] == nums[i-1]:
                count += 1
            else:
                count = 1
                
            if count <= 2:
                nums[j] = nums[i]
                j+=1
        print("length is: ",j)        
        return j